### PR TITLE
C#: Exclude Equals and GetHashCode overrides from model generation.

### PR DESCRIPTION
**Pull request by michaelnebel**

### CODE REVIEW

File: csharp/ql/src/utils/model-generator/internal/CaptureModelsSpecific.qll [hunk 1]
```diff
@@ -16,14 +17,25 @@ module TaintTracking = CS::TaintTracking;
 
 class Type = CS::Type;
 
+/**
+ * Holds if `api` is an override of `GetHashCode` or `Equals`.
+ */
+private predicate isIrrellevantObjectOveride(CS::Callable api) {
```
Comment (tamasvajk) [inline csharp/ql/src/utils/model-generator/internal/CaptureModelsSpecific.qll:23]: There's a typo in "Irrellevant" vs "Irrelevant".

**Review by tamasvajk — CHANGES_REQUESTED** [review]
Looks good apart from the minor typo.

File: csharp/ql/src/utils/model-generator/internal/CaptureModelsSpecific.qll [hunk 2]
```diff
@@ -16,14 +17,25 @@ module TaintTracking = CS::TaintTracking;
 
 class Type = CS::Type;
 
+/**
+ * Holds if `api` is an override of `GetHashCode` or `Equals`.
+ */
+private predicate isIrrelevantObjectOveride(CS::Callable api) {
```
Comment (tamasvajk) [inline csharp/ql/src/utils/model-generator/internal/CaptureModelsSpecific.qll:23]: Ahh, sorry, I didn't notice earlier, but "Overide" also has a typo.

File: csharp/ql/src/utils/model-generator/internal/CaptureModelsSpecific.qll [hunk 3]
```diff
@@ -16,14 +17,25 @@ module TaintTracking = CS::TaintTracking;
...
+  exists(System::SystemObjectClass c |
+    api = c.getGetHashCodeMethod().getAnOverrider*() or
+    api = c.getEqualsMethod().getAnOverrider*()
+  )
```
Comment (hvitved) [inline csharp/ql/src/utils/model-generator/internal/CaptureModelsSpecific.qll:27]: In order to make this more robust towards future additions (like [`IEquatable<T>.Equals(T)`](https://docs.microsoft.com/en-us/dotnet/api/system.iequatable-1.equals?view=net-6.0)), I would rewrite it slightly as
```ql
exists(Callable exclude |
  api = exclude.getAnOverrider*().getUnboundDeclaration() or
  api = exclude.getAnUltimateImplementor().getUnboundDeclaration()
|
  exists(System::SystemObjectClass c |
    exlude = [c.getGetHashCodeMethod(), c.getEqualsMethod()]
  )
)
```

File: csharp/ql/src/utils/model-generator/internal/CaptureModelsSpecific.qll (hunk 2 above)
Comment (michaelnebel) [reply csharp/ql/src/utils/model-generator/internal/CaptureModelsSpecific.qll:23]: OMG - TGIF! Thank you 👍

File: csharp/ql/src/utils/model-generator/internal/CaptureModelsSpecific.qll (hunk 3 above)
Comment (michaelnebel) [reply csharp/ql/src/utils/model-generator/internal/CaptureModelsSpecific.qll:28]: Yes, it is a good idea to also exclude IEquatable<T>.Equals.

**Review by hvitved — APPROVED** [review]

**Comment by michaelnebel** [comment]
The DCA doesn't indicate that there are any performance regressions.

**Review by hvitved — APPROVED** [review]

**Comment by michaelnebel** [comment]
> The DCA doesn't indicate that there are any performance regressions.

Pr. the discussion on slack in codeql-dist-compare. I loooked that actual workflow execution (and query execution times) and not only the overview page as the uinterpreted queries are not included in the overall analysis time.

**Pull request merged**